docs: ADR 047: Extend Upgrade Plan - Initial Draft (#10602)
cosmos/cosmos-sdk: docs/architecture/adr-047-extend-upgrade-plan.md

    Once complete, the exit code will be logged.

## Consequences

### Backwards Compatibility

Since the only change to existing definitions is the addition of the `upgrade` field to the `Plan` message, and that field is optional, there are no backwards incompatibilities with respects to the proto messages.
Additionally, current behavior will be maintained when no `UpgradeInstructions` are provided, so there are no backwards incompatibilities with respects to either the upgrade module or Cosmovisor.

### Forwards Compatibility

In order to utilize the `UpgradeInstructions` as part of a software upgrade, both of the following must be true:
1.  The chain must already be using a sufficiently advanced version of the Cosmos SDK.
1.  The chain's nodes must be using a sufficiently advanced version of Cosmovisor.

### Positive

1.  The structure for defining artifacts is clearer since it is now defined in the proto instead of in documentation.
1.  Availability of a pre-run command becomes more obvious.
1.  A post-run command becomes possible.

### Negative

1.  The `Plan` message becomes larger.
1.  There is no option for providing a URL that will return the `UpgradeInstructions`.

> 1.  The only way to provide multiple assets for a platform is to use an archive as the platform's artifact.

Reviewer: ```suggestion
1.  The only way to provide multiple assets (executable and other files) for a platform is to use an archive as the platform's artifact.
```

1.  A chain software upgrade is required before this new functionality can be used.

>     I.e. The next upgrade must use the old definitions, but the one after that can use the new structure.

Reviewer: I'm not sure if this is negative consequence: every feature which touches the state changes the state machines, so it's _breaking_.

Author: I included this as a negative because it's possible to design the Cosmovisor changes so that this isn't the case.

Specifically, update Cosmovisor to further overload the `info` field to allow either the JSON of this new `UpgradeInstructions` or a url that returns that JSON (as well as what's currently looked for there). Then, as long as the node is running the new version of Cosmovisor, the new functionality from ADR is available on the next chain upgrade.

I will add that to the "Cosmovisor Updates" section and then remove this as a negative.

Author: Removed this negative and added to the "Cosmovisor Updates" section, lines 134-136.

## Status

DRAFT Not Implemented

## Abstract

This ADR expands the existing upgrade `Plan` proto message to include new fields for defining pre-run and post-run processes.
It also defines a structure for providing downloadable assets involved in an upgrade.

## Context

The `upgrade` module in conjunction with Cosmovisor are designed to facilitate and automate a blockchain's transition from one version to another.

Users submit a software upgrade governance proposal containing an upgrade `Plan`.
The `Plan` currently contains the following fields:
- `name`: A short string identifying the new version.
- `height`: The chain height at which the upgrade is to be performed.
- `info`: A string containing information about the upgrade. It can either be a json object (with a specific structure defined through documentation), or it can be a URL that will return such json.

The json object from the `info` field identifies URLs used to download the new blockchain executable for different platforms (OS and Architecture, e.g. "linux/amd64").
Such a URL can either return the executable file directly or can return an archive containing the executable and possibly other assets.

> If an archive contains a `bin/pre-upgrade` file, Cosmovisor executes it (and waits for it to finish) before attempting to restart the chain.

Author: Ah. good catch. I was confusing `exec.Command(bin, "pre-upgrade")` with `filepath.Join`. Fixed.

Currently, there is no mechanism that makes Cosmovisor run a command after the upgraded chain has been restarted.

## Decision

### Protobuf Updates

We will define a new message for providing upgrade instructions and add it as a new field in the `Plan` message.
The upgrade instructions will contain a list of assets available for each platform.
It will also allow for the definition of a pre-run and post-run command.

```protobuf
message Plan {
  // ... (existing fields)

  UpgradeInstructions upgrade = 6;
}
```

The new `UpgradeInstructions upgrade` field MUST be optional.

```protobuf
message UpgradeInstructions {
  string pre_run = 1;
  string post_run = 2;

>   repeated Asset assets    = 3;

Author: I like `Asset assets` is better since they might be for archives with things other than executables.

  string description = 4;
}
```

All fields in the `UpgradeInstructions` SHOULD be optional.
- `pre_run` is a command to run prior to the upgraded chain restarting.
  If defined, this command SHOULD behave the same as the current [pre-upgrade](https://github.com/cosmos/cosmos-sdk/blob/master/docs/migrations/pre-upgrade.md) command.

>   The working directory this command runs from SHOULD be `{DAEMON_HOME}/{upgrade name}`.

Author: I've updated the `## Context` section with a little more about how downloads currently happen and what `DAEMON_HOME` is. I also corrected `{DAEMON_HOME}/{upgrade name}` to ``{DAEMON_HOME}/cosmovisor/{upgrade name}`.

- `post_run` is a command to run after the upgraded chain has been started. If defined, this command SHOULD be only executed once.
  The output and exit code SHOULD be logged but SHOULD NOT affect the running of the upgraded chain.
  The working directory this command runs from SHOULD be `{DAEMON_HOME}/{upgrade name}`.
- `assets` SHOULD allow any number of entries.
  It SHOULD have only one entry per platform.

> - `description` contains additional information about the upgrade and might contain references to external resources.

Author: Changed `additional` to `human-readable`. Also, the next line is `It SHOULD NOT be used for structured processing information.`

  It SHOULD NOT be used for structured processing information.

```protobuf
message Asset {
  string platform = 1;
  string url = 2;

>   string checksum = 3;

Author: Yeah. I defined it similar to how `getter` expects/handles it. It does make sense to split out the `checksum_algo` though.

I've added a `checksum_algo` field and things accordingly.

> }

Author: I'm not sure I like that idea. Again, we'd run into the question of what the working directory is for the commands, and it'd have to be different from the `pre_run` and `post_run` working directories. And since it'd have to be hard-coded, I'm not sure how useful it'd end up being.

Similar behavior could also be achieved just using the `pre_run` command. We'd just need to make sure that, after successfully downloading the asset, the `pre_run` command is run regardless of the existence of `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}`. That way, the asset for the platform can be an archive with a script. It'd be downloaded, then the `pre_run` command would execute that script which could then be responsible for downloading the new binary.

```

- `platform` is a required string that SHOULD be in the format `{OS}/{CPU}`, e.g. `"linux/amd64"`.
  The string `"any"` SHOULD also be allowed.
  An `Asset` with a `platform` of `"any"` SHOULD be used as a fallback when a specific `{OS}/{CPU}` entry is not found.
  That is, if an `Asset` exists with a `platform` that matches the system's OS and CPU, that should be used;
  otherwise, if an `Asset` exists with a `platform` of `any`, that should be used;
  otherwise no asset should be downloaded.

> - `url` is a required string that MUST conform to [RFC 1738: Uniform Resource Locators](https://www.ietf.org/rfc/rfc1738.txt).

Author: It's not necessarily a binary, but I've added a second sentence here to clarify what the URL is expected to return.

```text
- `url` is a required URL string that MUST conform to [RFC 1738: Uniform Resource Locators](https://www.ietf.org/rfc/rfc1738.txt).
  A request to this `url` MUST return either an executable file or an archive containing either `bin/{DAEMON_NAME}` or `{DAEMON_NAME}`.
```

- `checksum` SHOULD be provided, but is not required.

>   It SHOULD be a hex encoded checksum of the expected result of a request to the `url`.

Author: It's not required to be `sha256`. It could also be `sha512`, `sha1`, or `md5`.

I've updated this part a bit due to other recommendations.

    1.  If the `checksum` field has a value and the `url` also has a `checksum` query parameter, the two values MUST be equal.
    1.  The `url` MUST return either a file or an archive containing eitehr `bin/{DAEMON_NAME}` or `{DAEMON_NAME}`.
    1.  If a `checksum` is provided (in the field or as a query param), the checksum of the result of the `url` MUST equal the provided checksum.

### Cosmovisor updates

If the `upgrade-info.json` file does not contain any `UpgradeInstructions`, existing functionality will be maintained.

We will update Cosmovisor to look for and handle the new `UpgradeInstructions` in `upgrade-info.json`.
If the `UpgradeInstructions` are provided, we will do the following:
1.  The `info` field will be ignored.
1.  The `assets` field will be used to identify the asset to download based on the `platform` that Cosmovisor is running in.
1.  If the downloaded asset is an archive that contains a `bin/pre-upgrade` file, that file will be ignored.
1.  If a `checksum` is provided (either in the field or as a query param in the `url`), and the downloaded asset has a different checksum, the upgrade process will be interrupted and Cosmovisor will exit with an error.

> 1.  If a `pre_run` command is defined, it will be executed at the same point in the process where the `bin/pre-upgrade` file would have been executed.

Author: In this case, `bin/pre-upgrade` is the specific file currently looked for and used.

### Forwards Compatibility

In order to utilize the `UpgradeInstructions` as part of a software upgrade, both of the following must be true:
1.  The chain must already be using a sufficiently advanced version of the Cosmos SDK.
1.  The chain's nodes must be using a sufficiently advanced version of Cosmovisor.

### Positive

1.  The structure for defining downloadable assets is clearer since it is now defined in the proto instead of in documentation.
1.  Availability of a pre-run command becomes more obvious.
1.  A post-run command becomes possible.

### Negative

1.  The `Plan` message becomes larger.
1.  There is no option for providing a URL that will return the `UpgradeInstructions`.
    Existing functionality of the `info` field will remain, but will not allow for the newly defined structure.
1.  There is no built-in mechanism for preventing the `assets` from having two entries with the same `platform`.
    Behavior of this situation is undefined.
1.  All `assets` are assumed to either be the executable or be an archive that contains the executable.

>     There is no way to define more than one downloadable asset for a platform.

Author: An archive is just be one downloadable asset. It's true that the archive can have multiple files, but it's still just a single downloadable asset and is the only way of getting multiple files/executables.

# ADR 047: Extend Upgrade Plan

## Changelog

- Nov, 23, 2021: Initial Draft

## Status

> DRAFT Not Implemented

Author: Done.

## Abstract

> This ADR expands the existing x/upgrade `Plan` proto message to include new fields for defining pre-run and post-run processes.

Reviewer 2: ```suggestion
This ADR expands the existing x/upgrade `Plan` proto message to include new fields for defining pre-run and post-run processes within upgrade tooling.
```

Author: Change made.

It also defines a structure for providing downloadable artifacts involved in an upgrade.

## Context

The `upgrade` module in conjunction with Cosmovisor are designed to facilitate and automate a blockchain's transition from one version to another.

Users submit a software upgrade governance proposal containing an upgrade `Plan`.
The [Plan](https://github.com/cosmos/cosmos-sdk/blob/v0.44.5/proto/cosmos/upgrade/v1beta1/upgrade.proto#L12) currently contains the following fields:
- `name`: A short string identifying the new version.
- `height`: The chain height at which the upgrade is to be performed.
- `info`: A string containing information about the upgrade.

The `info` string can be anything.

> However, Cosmovisor will try to use the `info` field to automatically download a new version of the blockchain executable.

Reviewer 2: ```suggestion
The `info` string can be anything, however it is intended to specify the version of the blockchain executable to auto-download.
The name `info` is kept for backwards compatibility purposes.
```

Author: This `Context` section is just about how things currently are. It's not the place for the backwards compatibility note.
I'm also uncomfortable labeling the intent of the `info` field since it's currently not well defined (and unclear). Use of the field by Cosmovisor is defined, but the spec on the field in general isn't.

> However, Cosmovisor will try to use the `info` field to automatically download a new version of the blockchain executable.

Author: This change isn't entirely accurate. If a node owner has manually downloaded and placed the new version, then the `info` field is ignored. Cosmovisor will just switch to the new version.

I've added this line a little below:
> If an upgrade height is reached and the new version of the executable version isn't available, Cosmovisor will stop running.

The JSON object identifies URLs used to download the new blockchain executable for different platforms (OS and Architecture, e.g. "linux/amd64").
Such a URL can either return the executable file directly or can return an archive containing the executable and possibly other assets.

If the URL returns an archive, it is decompressed into `{DAEMON_HOME}/cosmovisor/{upgrade name}`.
Then, if `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}` does not exist, but `{DAEMON_HOME}/cosmovisor/{upgrade name}/{DAEMON_NAME}` does, the latter is copied to the former.
If the URL returns something other than an archive, it is downloaded to `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}`.

Both `DAEMON_HOME` and `DAEMON_NAME` are [environment variables used to configure Cosmovisor](https://github.com/cosmos/cosmos-sdk/blob/cosmovisor/v1.0.0/cosmovisor/README.md#command-line-arguments-and-environment-variables).

Currently, there is no mechanism that makes Cosmovisor run a command after the upgraded chain has been restarted.

## Decision

### Protobuf Updates

> We will define a new message for providing upgrade instructions and add it as a new field in the `Plan` message.

Author: Changed to this:
> We will update the `x/upgrade.Plan` message for providing upgrade instructions.

The upgrade instructions will contain a list of artifacts available for each platform.

> It will also allow for the definition of a pre-run and post-run command.

Reviewer 2: ```suggestion
It allows for the definition of a pre-run and post-run commands.
These commands are not consensus guaranteed, but emitted into a file that cosmosvisor will run during its upgrade handling. (And similar tooling can choose to run)
```

Author: Future tense is correct there, so I'm keeping that. I added something similar to the second sentence: 
> These commands are not consensus guaranteed; they will be executed by cosmosvisor (or other) during its upgrade handling.

> It will also allow for the definition of a pre-run and post-run command.

Author: Change made.

```protobuf
message Plan {
  // ... (existing fields)

>   UpgradeInstructions upgrade = 6;

Reviewer: ```suggestion
  UpgradeInstructions instructions = 6;
```

Author: Change made. Also updated a couple other places referring to this field.

}
```

The new `UpgradeInstructions upgrade` field MUST be optional.

```protobuf
message UpgradeInstructions {
  string pre_run              = 1;
  string post_run             = 2;
  repeated Artifact artifacts = 3;
  string description          = 4;
}
```

All fields in the `UpgradeInstructions` are optional.

> - `pre_run` is a command to run prior to the upgraded chain restarting.

Reviewer 3: does this mean, that the chain is still running the prior version of the code or has the chain stopped.

Reviewer: The chain stops for the update. So the process is:
* at the upgrade height stop the chain
* download new binaries if needed (with the upcoming updates it should be done even before)
* make pre-run 
* run the new app

Author: I updated this part a bit to better indicate the timing.

Reviewer 4: if users choose to backup data when upgrade, this happens before or after `pre-upgrade` ?

  If defined, then the app supervisors (e.g. Cosmovisor) MUST NOT run `app pre-run`.

>   If defined, this command MUST behave the same as the current [pre-upgrade](https://github.com/cosmos/cosmos-sdk/blob/v0.44.5/docs/migrations/pre-upgrade.md) command.

Reviewer: let's copy the pre-upgrade spec here.

  The working directory this command runs from MUST be `{DAEMON_HOME}/cosmovisor/{upgrade name}`.

> - `post_run` is a command to run after the upgraded chain has been started. If defined, this command MUST be only executed once.

Reviewer 3: how do we expect to be able to enforce only-once semantics?

Author: I updated a few spots on this to clarify that it's once per node (not once per chain).

  The output and exit code SHOULD be logged but SHOULD NOT affect the running of the upgraded chain.
  The working directory this command runs from MUST be `{DAEMON_HOME}/cosmovisor/{upgrade name}`.
- `artifacts` define items to be downloaded.
  It SHOULD have only one entry per platform.
- `description` contains human-readable information about the upgrade and might contain references to external resources.
  It SHOULD NOT be used for structured processing information.

```protobuf
message Artifact {
  string platform      = 1;
  string url           = 2;
  string checksum      = 3;
  string checksum_algo = 4;
}
```

- `platform` is a required string that SHOULD be in the format `{OS}/{CPU}`, e.g. `"linux/amd64"`.
  The string `"any"` SHOULD also be allowed.
  An `Artifact` with a `platform` of `"any"` SHOULD be used as a fallback when a specific `{OS}/{CPU}` entry is not found.
  That is, if an `Artifact` exists with a `platform` that matches the system's OS and CPU, that should be used;
  otherwise, if an `Artifact` exists with a `platform` of `any`, that should be used;

>   otherwise no artifact should be downloaded.

Reviewer: Maybe we should return an error if there are artifacts, but none matches the running platform?

Reviewer: So, if there is a common, shared artifact, then it has to be copied to all platforms? Let's make a note about that.

Author: This description is what Cosmovisor currently does. If this new stuff is going to have different behavior, we should use a word other than `any` for it.

I think current behavior is sufficient, though. We can maybe update Cosmovisor to issue a warning if there are multiple platforms, but the desired one isn't found and "any" is used instead, but I don't think that behavior needs to be part of spec.

- `url` is a required URL string that MUST conform to [RFC 1738: Uniform Resource Locators](https://www.ietf.org/rfc/rfc1738.txt).

>   A request to this `url` MUST return either an executable file or an archive containing either `bin/{DAEMON_NAME}` or `{DAEMON_NAME}`.

Author: Change made.

        1. The `checksum` query parameter value MUST be in the format of `{checksum_algo}:{checksum}`.
        1. The `{checksum}` from the query parameter MUST equal the `checksum` provided in the `Artifact`.
        1. The `{checksum_algo}` from the query parameter MUST equal the `checksum_algo` provided in the `Artifact`.
1.  The following validation is currently done using the `info` field. We will apply similar validation to the `UpgradeInstructions`.
    For each `Artifact`:
    1.  The `platform` MUST have the format `{OS}/{CPU}` or be `"any"`.
    1.  The `url` field MUST NOT be empty.
    1.  The `url` field MUST be a proper URL.
    1.  A `checksum` MUST be provided either in the `checksum` field or as a query parameter in the `url`.
    1.  If the `checksum` field has a value and the `url` also has a `checksum` query parameter, the two values MUST be equal.
    1.  The `url` MUST return either a file or an archive containing either `bin/{DAEMON_NAME}` or `{DAEMON_NAME}`.
    1.  If a `checksum` is provided (in the field or as a query param), the checksum of the result of the `url` MUST equal the provided checksum.

> 

Reviewer: ```suggestion
1. `Plan.Info` will be deprecated - tools and CLI should return a warning if `Plan.Info` is provided but `UpgradeInstructions` is not.

```

Author: I've held off on making this change because I'd still like to have a way to provide a URL to JSON of upgrade instructions.

If we want to fully deprecate the `info` field, then we'd need to add a new field specifically for such a URL.

We could also use a `one of` for the `instructions`, to make it clear that you either define the instructions in the message or provide a url to them. But I'm not a fan of the `one of` stuff.

Author: This other conversation was similar to this one. I "resolved" that one and linked to this one.

https://github.com/cosmos/cosmos-sdk/pull/10602#discussion_r782472640

Author: Sorry to chain this further, but this discussion is also going on here https://github.com/cosmos/cosmos-sdk/pull/10602#discussion_r781438288. I will close this one and continue it there.

Downloading of an `Artifact` will happen the same way that URLs from `info` are currently downloaded.

### Cosmovisor Updates

> If the `upgrade-info.json` file does not contain any `UpgradeInstructions`, existing functionality will be maintained.

Reviewer: ```suggestion
Current meaning of `Plan.Info` will be deprecated - the attribute will continue to exist, but it will have a non machine interpretable content.
    * If the `upgrade-info.json` file contains `UpgradeInstructions` then `Plan.Info` won't be used.
    *  otherwise, hypervisors are expected to fallback to the existing behavior and try to interpret `Plan.Info`.
```

Author: I'm going to resolve this conversation so it's not happening in two places.

Here's the other conversation: https://github.com/cosmos/cosmos-sdk/pull/10602#discussion_r782472640

We will update Cosmovisor to look for and handle the new `UpgradeInstructions` in `upgrade-info.json`.
If the `UpgradeInstructions` are provided, we will do the following:
1.  The `info` field will be ignored.
1.  The `artifacts` field will be used to identify the artifact to download based on the `platform` that Cosmovisor is running in.
1.  If a `checksum` is provided (either in the field or as a query param in the `url`), and the downloaded artifact has a different checksum, the upgrade process will be interrupted and Cosmovisor will exit with an error.
1.  If a `pre_run` command is defined, it will be executed at the same point in the process where the `app pre-upgrade` command would have been executed.
    It will be executed using the same environment as other commands run by Cosmovisor.
1.  If a `post_run` command is defined, it will be executed after executing the command that restarts the chain.
    It will be executed in a background process using the same environment as the other commands.
    Any output generated by the command will be logged.
    Once complete, the exit code will be logged.

> We will update Cosmovisor to allow JSON of the new `UpgradeInstructions` to be allowed in the existing `info` field.

Reviewer: why? I think we shouldn't mix the two any more. They should co-exist:
```
{
  "info": "some text"
  "instructions: { object }
}
```

Author: Being able to retroactively change an upgrade plan is sometimes desirable. Allowing a url with `instructions` JSON to be provided allows such a change. This is important in instances when security vulnerabilities are found and fixed and you don't want nodes to ever try using the vulnerable version.

I've updated the sentence to:
> We will update Cosmovisor to allow the `info` field to be a URL to JSON of the new `UpgradeInstructions` structure.

Author: Another option that would handle this situation very well is to add a `signer` next to the `checksum`. If a `signer` is provided, then the asset must be signed by that signer. It's not expected that someone will want to use a `checksum` and `signer` but I don't think we should disallow it.

By having a `signer`, it would be possible to securely change an asset without having to issue a new upgrade proposal.

Reviewer: But why putting this in `info`? I wouldn't touch it (except deprecating it's meaning). Maybe we can create a new attribute: `Plan.InstructionsURL`. And we should make a validation that both `Plan.instructions` or `Plan.instructionsURL` can't be defined at the same time.

Author: I held off on adding a `Plan.InstructionsURL` field since that functionality already exists on the info field. All that's changing is the addition of a new structure option. Additionally, we'd want to also allow definition of a `checksum` and/or `signer` (like the other asset urls). The `Asset` message could be used here except that a `platform` isn't desirable in this case. Comments could be used to indicate that though. But the security available for the other assets should also be available for the instructions url.

For this type of thing to work securely, we really need to add a `signer` field to the `Asset` message. However, I know that that adds a considerable amount of work to implement. I thought it might be best to have discussions on it first.

Author: I resolved this discussion: https://github.com/cosmos/cosmos-sdk/pull/10602#discussion_r789032275 in order to continue it here.

After discussing this with the group, we decided to not do a `signer` field and to also deprecate the url/json in the `info` field. We also decided not to include an `InstructionsURL` field. I will update the doc to reflect this, then resolve this discussion.

### Reasoning

The chain only ever has at most one upgrade plan. If a chain has a version that shouldn't be used (e.g. security vulnerability), they would put in a new upgrade plan with a fixed version. If the bad version isn't yet in use, the new plan will overwrite the old plan and the bad version will be ignored. If the bad version is already running, then the new upgrade plan will update the nodes as usual.

The functionality of a `signer` is needed when an asset needs to change after the upgrade plan has been submitted. If the instructions can no longer be an asset, the only things that would be changing are the executables (or archives with the executables). If those change, though, then a new upgrade plan should be submitted anyway.

When parsing the `info` field, Cosmovisor will first look for the new `UpgradeInstructions` structure; if not found, Cosmovisor will fallback to existing functionality.

## Consequences

### Backwards Compatibility

Since the only change to existing definitions is the addition of the `upgrade` field to the `Plan` message, and that field is optional, there are no backwards incompatibilities with respects to the proto messages.
Additionally, current behavior will be maintained when no `UpgradeInstructions` are provided, so there are no backwards incompatibilities with respects to either the upgrade module or Cosmovisor.

### Forwards Compatibility

In order to utilize the `UpgradeInstructions` as part of a software upgrade, both of the following must be true:
1.  The chain must already be using a sufficiently advanced version of the Cosmos SDK.
1.  The chain's nodes must be using a sufficiently advanced version of Cosmovisor.

### Positive

1.  The structure for defining artifacts is clearer since it is now defined in the proto instead of in documentation.
1.  Availability of a pre-run command becomes more obvious.
1.  A post-run command becomes possible.

### Negative

> 1.  The `Plan` message becomes larger.

Reviewer: ```suggestion
1.  The `Plan` message becomes larger. However this won't be a problem, because proposals require a bond, and they are pruned.
```

Author: Proposals do require a bond, but they also cost gas which is affected by message size. And pruning depends on node setup, so it's not accurate to say definitively that they're pruned.

Reviewer: Prunning doesn't depend on the node setup. Inactive or passed proposals are pruned.

Author: Updated to this:
```md
1.  The `Plan` message becomes larger. This is negligable because A) the `x/upgrades` module only stores at most one upgrade plan, and B) upgrades are rare enough that the increased gas cost isn't a concern.
```

- `height`: The chain height at which the upgrade is to be performed.
- `info`: A string containing information about the upgrade.

The `info` string can be anything.
However, Cosmovisor will try to use the `info` field to automatically download a new version of the blockchain executable.
For the auto-download to work, Cosmovisor expects it to be either a stringified JSON object (with a specific structure defined through documentation), or a URL that will return such JSON.
The JSON object identifies URLs used to download the new blockchain executable for different platforms (OS and Architecture, e.g. "linux/amd64").
Such a URL can either return the executable file directly or can return an archive containing the executable and possibly other assets.

If the URL returns an archive, it is decompressed into `{DAEMON_HOME}/cosmovisor/{upgrade name}`.
Then, if `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}` does not exist, but `{DAEMON_HOME}/cosmovisor/{upgrade name}/{DAEMON_NAME}` does, the latter is copied to the former.
If the URL returns something other than an archive, it is downloaded to `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}`.

> If an upgrade height is reached and the new version of the executable version isn't available, Cosmovisor will stop running.

Author: Yes. There are several reasons this might happen, and Cosmovisor outputs different errors for each before stopping.

Possible reasons:
1. There is no url for the given architecture:
2. The url for the given architecture does not return a file or archive.
3. A checksum was provided for the asset that doesn't match the asset retrieved from the URL.
4. The asset is an archive and a `$DAEMON_NAME` file cannot be found in either `/bin` or `/`.
5. The `/bin/$DAEMON_NAME` file exists but is not executable and cannot be made executable. This might happen if a node operator manually downloads and places the new file without making it executable, and Cosmovisor doesn't have the permissions needed to fix it.

  This command MUST behave the same as the current [pre-upgrade](https://github.com/cosmos/cosmos-sdk/blob/v0.44.5/docs/migrations/pre-upgrade.md) command.
  It does not take in any command-line arguments and is expected to terminate with the following exit codes:

  | Exit status code | How it is handled in Cosmosvisor                                                                                    |
  |------------------|---------------------------------------------------------------------------------------------------------------------|
  | `0`              | Assumes `pre-upgrade` command executed successfully and continues the upgrade.                                      |
  | `1`              | Default exit code when `pre-upgrade` command has not been implemented.                                              |
  | `30`             | `pre-upgrade` command was executed but failed. This fails the entire upgrade.                                       |
  | `31`             | `pre-upgrade` command was executed but failed. But the command is retried until exit code `1` or `30` are returned. |
  If defined, then the app supervisors (e.g. Cosmovisor) MUST NOT run `app pre-run`.

> - `post_run` is a command to run after the upgraded chain has been started. If defined, this command MUST be only executed at most once by an upgrading node.

Author: I'll add an upgrade timeline part to one of the sections to explain timing, but it goes like this:

1. Chain reaches upgrade height, writes the upgrade json file and halts.
2. Cosmovisor backs up the data (if set up to do so).
3. Cosmovisor downloads the new version.
3. The `pre_run` command is run.
4. The chain restarts with the new version.
5. Immediately after issuing the command to restart the chain, the `post_run` command is run in a new, detached process.

1. Cosmovisor executes **the `pre_run` command if provided**, or else the `${DAEMON_NAME} pre-upgrade` command.
1. Cosmovisor restarts the app using the new version and same args originally provided.
1. **Cosmovisor immediately runs the `post_run` command in a detached process.**

## Consequences

### Backwards Compatibility

Since the only change to existing definitions is the addition of the `instructions` field to the `Plan` message, and that field is optional, there are no backwards incompatibilities with respects to the proto messages.
Additionally, current behavior will be maintained when no `UpgradeInstructions` are provided, so there are no backwards incompatibilities with respects to either the upgrade module or Cosmovisor.

### Forwards Compatibility

In order to utilize the `UpgradeInstructions` as part of a software upgrade, both of the following must be true:
1.  The chain must already be using a sufficiently advanced version of the Cosmos SDK.
1.  The chain's nodes must be using a sufficiently advanced version of Cosmovisor.

### Positive

1.  The structure for defining artifacts is clearer since it is now defined in the proto instead of in documentation.

> 1.  Availability of a pre-run command becomes more obvious.
> 1.  A post-run command becomes possible.

Reviewer 2: ```suggestion
1.  Availability of a pre-run command becomes more obvious.
1.  A post-run command becomes possible.
1.  Node operators can easily override these
```
